src/bin/electrs.rs

```diff
@@ -123,14 +125,33 @@ fn run_server(config: Arc<Config>) -> Result<()> {
         "count of iterations of electrs main loop each 5 seconds or after interrupts",
     ));
 
-    loop {
+    'outer: loop {
         main_loop_count.inc();
 
-        if let Err(err) = signal.wait(Duration::from_secs(5), true) {
-            info!("stopping server: {}", err);
-            rest_server.stop();
-            // the electrum server is stopped when dropped
-            break;
+        // In the next for loop:
+        // We are going to wait 5 secs (50*100ms) if nothings happens.
+        // We are stopping if we receive a TERM or INT signal.
+        // We are interrupting the wait if we receive a block hash notification (if zmq enabled) or receiving a USR1 signal.
+        for _ in 0..50 {
```

Comment:
Perhaps we could use crossbeam_channel's [`select!`](https://docs.rs/crossbeam-channel/latest/crossbeam_channel/macro.select.html) to wait for the first receiver that emits, instead of polling `block_hash_receive` on an interval?

Also, this could be made part of `Waiter::wait()`, with `accept_sigusr` changed into `accept_new_block_notif` to cover both cases.

Reply:
Changed to use `select!` in latest force push which is also rebased.
Note that since we want to use `select!` I removed previous latest commit, introduced `select!` so that one branch is the timeout, then added the zmq branch